src/marqo/core/search/hybrid_search.py

```diff
@@ -0,0 +1,212 @@
+            ef_search=ef_search,
+            approximate=approximate,
+            offset=offset,
+            or_phrases=optional_terms,
+            and_phrases=required_terms,
+            attributes_to_retrieve=attributes_to_retrieve,
+            searchable_attributes=searchable_attributes,
+            score_modifiers=score_modifiers.to_marqo_score_modifiers() if score_modifiers is not None else None,
+            # Hybrid-specific attributes
+            score_modifiers_lexical=hybrid_parameters.score_modifiers_lexical.to_marqo_score_modifiers()
+            if hybrid_parameters.score_modifiers_lexical is not None else None,
+            score_modifiers_tensor=hybrid_parameters.score_modifiers_tensor.to_marqo_score_modifiers()
+            if hybrid_parameters.score_modifiers_tensor is not None else None,
+            hybrid_parameters=hybrid_parameters
+        )
+
+        vespa_index = vespa_index_factory(marqo_index)
+        vespa_query = vespa_index.to_vespa_query(marqo_query)
+
+        total_preprocess_time = RequestMetricsStore.for_request().stop("search.hybrid.processing_before_vespa")
+        logger.debug(
+            f"search (hybrid) pre-processing: took {(total_preprocess_time):.3f}ms to vectorize and process query.")
+
+        # SEARCH TIMER-LOGGER (roundtrip)
+        with RequestMetricsStore.for_request().time("search.hybrid.vespa",
+                                                    lambda t: logger.debug(f"Vespa search: took {t:.3f}ms")
+                                                    ):
+            try:
+                responses = config.vespa_client.query(**vespa_query)
+            except VespaStatusError as e:
+                # The index will not have the embedding_similarity rank profile if there are no tensor fields
+                if f"No profile named '{RANK_PROFILE_HYBRID_CUSTOM_SEARCHER}'" in e.message:
+                    raise core_exceptions.InvalidArgumentError(
+                        f"Index {index_name} either has no tensor fields or no lexically searchable fields, "
+                        f"thus hybrid search cannot be performed. "
+                        f"Please create an index with both tensor and lexical fields, or try a different search method."
+                    )
+                raise e
+
+        if not approximate and (responses.root.coverage.coverage < 100 or responses.root.coverage.degraded is not None):
+            raise errors.InternalError(
+                f'Graceful degradation detected for non-approximate search. '
+                f'Coverage is not 100%: {responses.root.coverage}'
+                f'Degraded: {str(responses.root.coverage.degraded)}'
+            )
+
+        # SEARCH TIMER-LOGGER (post-processing)
+        RequestMetricsStore.for_request().start("search.hybrid.postprocess")
+        gathered_docs = gather_documents_from_response(responses, marqo_index, highlights, attributes_to_retrieve)
+
+        if boost is not None:
```

Comment:
Move this validation up so we don't need to do search if input is invalid/ not supported.

Reply:
done